### PR TITLE
Cavatica Proxy

### PR DESCRIPTION
**Pull request by joneubank**

### CODE REVIEW

File: cavatica-proxy/src/main/java/io/kidsfirst/keys/cavatica/proxy/CavaticaProxy.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+      out.writeBytes(body);
+      out.flush();
+      out.close();
+    }
+
+    int status = con.getResponseCode();
+
+    BufferedReader reader = new BufferedReader(new InputStreamReader(con.getInputStream()));
+    StringBuffer content = new StringBuffer();
+    reader.lines().forEach(line->content.append(line));
```
Comment (andricDu) [inline cavatica-proxy/src/main/java/io/kidsfirst/keys/cavatica/proxy/CavaticaProxy.java]: Reference that method :musical_note: 

```java
reader.lines().forEach(content::append);
```

File: cavatica-proxy/src/main/java/io/kidsfirst/keys/cavatica/proxy/CavaticaProxy.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      return KMSUtils.decrypt(secretValue);
+
+    } else {
+      throw new IllegalArgumentException("No cavatica secret found, unable to send request.");
+    }
+  }
+
+
+  private String sendCavaticaRequest(String cavaticaKey, String path, String method, String body)
+    throws URISyntaxException, IOException {
```
Comment (andricDu) [inline cavatica-proxy/src/main/java/io/kidsfirst/keys/cavatica/proxy/CavaticaProxy.java]: `URISyntaxException` seems to be not needed.

File: cavatica-proxy/src/main/java/io/kidsfirst/keys/cavatica/proxy/CavaticaProxy.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+  }
+
+  private String getCavaticaKey(String userId) throws IllegalArgumentException {
+    List<Secret> allSecrets = SecretDao.getSecret("cavatica", userId);
+
+    if (!allSecrets.isEmpty()) {
+      Secret secret = allSecrets.get(0);
+
```
Comment (andricDu) [inline cavatica-proxy/src/main/java/io/kidsfirst/keys/cavatica/proxy/CavaticaProxy.java]: delete `\n`

File: cavatica-proxy/src/main/java/io/kidsfirst/keys/cavatica/proxy/CavaticaProxy.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+      String secretValue = secret.getSecret();
+
```
Comment (andricDu) [inline cavatica-proxy/src/main/java/io/kidsfirst/keys/cavatica/proxy/CavaticaProxy.java]: delete `\n`

File: cavatica-proxy/src/main/java/io/kidsfirst/keys/cavatica/proxy/CavaticaProxy.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+      return KMSUtils.decrypt(secretValue);
+
```
Comment (andricDu) [inline cavatica-proxy/src/main/java/io/kidsfirst/keys/cavatica/proxy/CavaticaProxy.java]: delete `\n`

File: cavatica-proxy/src/main/java/io/kidsfirst/keys/cavatica/proxy/CavaticaProxy.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+      if ( Arrays.stream(HTTP_ALLOWED_METHODS).noneMatch(allowed -> allowed.equals(method)) ) {
+        // Invalid method provided
+        throw new IllegalArgumentException("Provided method '"+method+"'is not allowed.");
+      }
+
+      // Body - default to null
+      JSONObject bodyJson = ((JSONObject) eventBody.get("body"));
+      String body = bodyJson == null ? null : bodyJson.toJSONString();
+
+      String response = sendCavaticaRequest(cavaticaKey, path, method, body);
```
Comment (andricDu) [inline cavatica-proxy/src/main/java/io/kidsfirst/keys/cavatica/proxy/CavaticaProxy.java]:
```java
return sendCavaticaRequest(cavaticaKey, path, method, body);
```

File: cavatica-proxy/src/main/java/io/kidsfirst/keys/cavatica/proxy/CavaticaProxy.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+      DataOutputStream out = new DataOutputStream(con.getOutputStream());
+      out.writeBytes(body);
+      out.flush();
+      out.close();
+    }
+
+    int status = con.getResponseCode();
+
+    BufferedReader reader = new BufferedReader(new InputStreamReader(con.getInputStream()));
+    StringBuffer content = new StringBuffer();
```
Comment (andricDu) [inline cavatica-proxy/src/main/java/io/kidsfirst/keys/cavatica/proxy/CavaticaProxy.java]: `StringBuffer` is used when you need thread safety as each method call uses synchronize. This means it is fairly slow which could be bad for the billing in lambda. 

Use `StringBuilder` instead.

**Comment by andricDu** [comment]
Excellent README!

**Comment by joneubank** [comment]
Updated. Good catch on StringBuffer vs StringBuilder.

**Pull request merged**